homeassistant/components/sensor/__init__.py

```diff
@@ -545,12 +553,17 @@ def from_dict(cls, restored: dict[str, Any]) -> SensorExtraStoredData | None:
                 native_value = dt_util.parse_datetime(native_value["isoformat"])
             elif type_ == "<class 'datetime.date'>":
                 native_value = dt_util.parse_date(native_value["isoformat"])
+            elif type_ == "<class 'decimal.Decimal'>":
+                native_value = Decimal(native_value["decimal_str"])
         except TypeError:
             # native_value is not a dict
             pass
         except KeyError:
             # native_value is a dict, but does not have all values
             return None
+        except DecimalInvalidOperation:
+            # native value couldn't be cast back to Decimal
+            return None
```

Comment:
Let's add a test for this case 👍

Reply:
while creating a test realised this can never happen, only chance would be for the storage to corrupt